cpp/src/link_test.cpp

```diff
@@ -23,9 +23,114 @@
 #include <proton/sender_options.hpp>
 #include <proton/receiver_options.hpp>
 #include <proton/container.hpp>
+#include <proton/connection.hpp>
+#include <proton/connection_options.hpp>
+#include <proton/listen_handler.hpp>
+#include <proton/listener.hpp>
+#include <proton/messaging_handler.hpp>
+#include <proton/types.hpp>
+#include <proton/message.hpp>
+#include <proton/target_options.hpp>
+#include <proton/source_options.hpp>
+#include <proton/delivery.hpp>
 
 #include <iostream>
 
+#include <map>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace {
+std::mutex m;
+std::condition_variable cv;
+bool listener_ready = false;
+int listener_port;
+} // namespace
+
+class test_recv : public proton::messaging_handler {
+  private:
+    class listener_ready_handler : public proton::listen_handler {
+        void on_open(proton::listener &l) override {
+            {
+                std::lock_guard<std::mutex> lk(m);
+                listener_port = l.port();
+                listener_ready = true;
+            }
+            cv.notify_one();
+        }
+    };
+
+    std::string url;
+    proton::listener listener;
+    listener_ready_handler listen_handler;
+
+  public:
+    test_recv(const std::string &s) : url(s) {}
+
+    void on_container_start(proton::container &c) override {
+        listener = c.listen(url, listen_handler);
+    }
+
+    void on_message(proton::delivery &d, proton::message &msg) override {
+        proton::symbol sym = "symbol";
+        proton::value val = "value";
+        std::map<proton::symbol, proton::value> props = d.receiver().target().dynamic_node_properties();
+
+        ASSERT(!props.empty());
+        for(auto it=props.begin(); it!=props.end(); it++) {
+            ASSERT_EQUAL(sym, it->first);
+            ASSERT_EQUAL(val, it->second);
+        }
+        d.receiver().close();
+        d.connection().close();
+        listener.stop();
+    }
+};
+
+class test_send : public proton::messaging_handler {
+  private:
+    std::string url;
+    proton::sender sender;
+
+  public:
+    test_send(const std::string &s) : url(s) {}
+
+    void on_container_start(proton::container &c) override {
+        proton::target_options opts;
+        std::map<proton::symbol,proton::value> m({{proton::symbol("symbol"), proton::value("value")}});
+        opts.dynamic_node_properties(m);
+        sender = c.open_sender(url, proton::sender_options().target(opts));
+    }
+
+    void on_sendable(proton::sender &s) override {
+        proton::message msg;
+        msg.body("message");
+        proton::tracker t = s.send(msg);
+        s.connection().close();
+    }
+};
+
+int test_dynamic_node_properties() {
```

Comment:
This test sets properties on a sender and then reads them out from the receiver's target. I am wondering if the other direction is also needed to be tested: set properties on receiver and read them from sender. Possibly not.

Reply:
Well, I am also wondering if the other direction needs to be tested as well or not. I gave it a try (set the properties on the receiver and try to read them from the sender in the same test).

- If I am setting dynamic node properties on the receiver like this `c.receiver_options(proton::receiver_options().source(opts));` and reading it from the sender then the test is getting terminated. And I can not see them getting set in the `attach frame`, plus the SOURCE in the third and fourth attach frame is NULL, which might be the reason for termination.

- But if I am setting it like this`c.receiver_options(proton::receiver_options().target(opts));` , then all is good.

The question is which one is correct or none of them is correct.

Meanwhile, I am looking for resources to get the idea.

Below are the attach frames:
```
[0x1fef6d0]: AMQP:FRAME:0 -> @attach(18) [name="3d0a3e34-a9fd-42b8-9cfd-0b1bbcc9c910", handle=0x0,
role=false, snd-settle-mode=0x2, rcv-settle-mode=0x0, source=@source(40) [address=null, durable=0x0, expiry-
policy=null, timeout=0x0, dynamic=false, dynamic-node-properties=null, distribution-mode=null, filter=null, default-
outcome=null, outcomes=null, capabilities=null], target=@target(41) [address="test", durable=0x0, expiry-policy=null,
 timeout=0x0, dynamic=false, dynamic-node-properties={:symbol="value"}, capabilities=null], unsettled=null, incomplete-
unsettled=null, initial-delivery-count=0x0, max-message-size=0x0, offered-capabilities=null, desired-capabilities=null,
 properties=null]
```
```
[0x7f7ffc002de0]: AMQP:FRAME:0 <- @attach(18) [name="3d0a3e34-a9fd-42b8-9cfd-0b1bbcc9c910", handle=0x0, 
role=false, snd-settle-mode=0x2, rcv-settle-mode=0x0, source=@source(40) [address=null, durable=0x0, expiry-
policy=null, timeout=0x0, dynamic=false, dynamic-node-properties=null, distribution-mode=null, filter=null, default-
outcome=null, outcomes=null, capabilities=null], target=@target(41) [address="test", durable=0x0, expiry-policy=null,
timeout=0x0, dynamic=false, dynamic-node-properties={:symbol="value"}, capabilities=null], unsettled=null, incomplete-
unsettled=null, initial-delivery-count=0x0, max-message-size=0x0, offered-capabilities=null, desired-capabilities=null,
properties=null]
```
```
[0x7f7ffc002de0]: AMQP:FRAME:0 -> @attach(18) [name="3d0a3e34-a9fd-42b8-9cfd-0b1bbcc9c910", handle=0x0, 
role=true, snd-settle-mode=0x2, rcv-settle-mode=0x0, source=null, target=@target(41) [address="test", durable=0x0,
expiry-policy=null, timeout=0x0, dynamic=false, dynamic-node-properties={:symbol2="value2"}, capabilities=null],
unsettled=null, incomplete-unsettled=null, initial-delivery-count=0x0, max-message-size=0x0, offered-capabilities=null, 
desired-capabilities=null, properties=null]
```
```
[0x1fef6d0]: AMQP:FRAME:0 <- @attach(18) [name="3d0a3e34-a9fd-42b8-9cfd-0b1bbcc9c910", handle=0x0, role=true,
 snd-settle-mode=0x2, rcv-settle-mode=0x0, source=null, target=@target(41) [address="test", durable=0x0, expiry-
policy=null, timeout=0x0, dynamic=false, dynamic-node-properties={:symbol2="value2"}, capabilities=null],
unsettled=null, incomplete-unsettled=null, initial-delivery-count=0x0, max-message-size=0x0, offered-capabilities=null, 
desired-capabilities=null, properties=null]
```
[Complete logs](https://gist.github.com/DreamPearl/f35d3431a28c0432ad04352c4b0bffa5)

Reply:
When being the 'requesting client' party asking for a consumer link to attach and have a dynamic node be created by setting the dynamic flag [and optionally setting dynamic node properties] that is done via the _source_, as it is requesting somewhere to consume from is dynamically created for it. The other side would read the source details requested by the requesting peer and action things, creating a node, giving it an address, and responding with its details. When it is sending its 'reply' attach it sets the live details that now exist also via the source (which from its perspective is now the local side, which it is authoritative on). The original peer can then read those details to establish the address and properties of the created node.

For a producer, it is similar but the _target_ is where the details will be set, as the original peer is requesting that somewhere to send to is dynamically created for it. The other peer will action this and respond with details via the _target_ to indicate what actually exists.

Reply:
A dynamic consumer is by far the primary use case for dynamic, so if only having one test it would be that (essentially the opposite of this test). Having one test for a consumer, and another for a producer (from the client perspective) wouldnt be a bad idea, they are separate things, but dynamic senders are rare.

The test doesnt actually do all the steps I'd say it should. Its possible another test is doing some of them somewhere, or that its just not tested for the cpp binding. It should use a 'dynamic' source/target and verify the behaviour in full, right now it isnt, but is instead attaching to a known address. The 'dynamic' handling involves the original peer not setting an address, and the other peer generating an address for the dynamic node and supplying its detail back to the original side in its reply attach, which the requesting peer can then read to determine the name of the dynamic node created for it. The dynamic-node-properties adds to that mechanism by letting the requester also ask for certain node details to be applied. The generating peer can always use that same field to indicate properties actually in effect for the dynamic node it created.

The test only looks to verify that the node properties were sent [illegally due to not setting dynamic] from requestor to the other side. It doesnt check it can handle reading the details that come back, i.e to verify any dynamic address was given, or that any dynamic-node-properties arrived.

As an example, here is a Java test verifying the dynamic flag and address propagation aspects (it doesnt test the dynamic node properties) for a dynamic consumer:
https://github.com/vert-x3/vertx-proton/blob/4.2.2/src/test/java/io/vertx/proton/ProtonClientTest.java#L468-L545

I dont know if the 'built in' handshaking bits in proton-cpp that are opening the link in your test will inspect the dynamic flag and generate an address by themselves for the attach. If they do the test could just at least verify there is a generated value. If not, or better still, the test could set a generated address at the server side and check the client side can pick it up when its told the link opened. It could do the same for the dynamic node properties, e.g have the server apply something other than originally asked for, verify they are picked up too at the client once the link opens.

Reply:
For now, I have the server apply dynamic node properties other than the desired properties asked by the client. And verifying that they are picked up by the client.

pn_log:  https://gist.github.com/DreamPearl/129d9d7d4f22b237269f6a5beefb227f

Reply:
That bit sounds/looks good. There are peripheral issues as mentioned before that should also be addressed. 

The trace log shows the test still uses an illegal target definition in the clients sender attach, where it sets dynamic-node-properties, but doesnt set dynamic=true, and sets an address "test" for the target (which the server naively copies and so reflects back). Thats then compounded when the test/client expects a 'response' attach with dynamic=true, something which shouldn't occur in this scenario.

Right now it verifies there are [different] dynamic node properties on the the servers 'response' attach, and the dynamic flag is set on the target, but not that there is a generated address.  It would be good for the test to verify the whole of the 'dynamic' interation, i.e that the client 'requests' dynamic=true and doesnt set an address, and that the client can pick up the address value set by the server once the link opens (since the dynamic node would be useless without an address).

I think it would be better if the server did its work in the 'on_receiver_open' (or per next comment, sender open) method rather than using the options in the container init, since you wont/cant generally use the options in that situation as a server as not everything is dynamic, and each dynamic node needs its own address etc. When the link attaches the server could verify it got a proper/expected request for a dynamic node, and respond setting the appropriate details on the target directly (or probably rather the source, per next comment), rather than using the options and also might as well open the link itself and not bothering with calling messaging_handler::on_receiver_open(r);).

Main other comment would be that this is still only trying to test a 'dynamic sender' from the clients perspective, when almost noone will ever do that, so it should either use a 'dynamic receiver' or test both.

Reply:
Although @gemmellr is correct in his analysis of what is going on in a real AMQP messaging interaction, this is probably not all that important as the C++ binding (at least currently) has no logic at either sender or receiver to do anything special for dynamic nodes - it just passes the info on to the application which is expected to validate and act on that information.
So from the point of view of this test it is only important to check that the properties get correctly carried in both directions. Unless the tester itself generates the requested nodes the API is not going to so there really isn't anything to validate in the test.
In a future change to the C++ binding it would definitely make sense to validate that you can't set address and set dynamic if you are sender/target or receiver/source. And also that if dynamic node properties are set the dynamic must also be set for the node, but that isn't done by the binding (or proton-c) currently so testing for it makes no sense. Producing realistic protocol interactions makes some sense but only to make the test "more realistic".